joint_limits/src/simple_joint_limiter.cpp

```diff
@@ -0,0 +1,393 @@
+            motion_after_stopping_duration)) ||
+          (desired_vel[index] > 0 &&
+           (joint_limits_[index].max_position - current_joint_states.positions[index] <
+            motion_after_stopping_duration)))
+        {
+          pos_limit_trig_jnts[index] = true;
+          braking_near_position_limit_triggered = true;
+        }
+        // else no need to slow down. in worse case we won't hit the limit at current velocity
+      }
+    }
+  }
+
+  // update variables according to triggers
+  if (braking_near_position_limit_triggered)
+  {
+    // this limit applies to all joints even if a single one is triggered
+    std::ostringstream ostr;
+    for (size_t index = 0; index < number_of_joints_; ++index)
+    {
+      // Compute accel to stop
+      // Here we aren't explicitly maximally decelerating, but for joints near their limits this
+      // should still result in max decel being used
+      desired_acc[index] = -current_joint_states.velocities[index] / dt_seconds;
+      if (joint_limits_[index].has_deceleration_limits)
+      {
+        desired_acc[index] = std::copysign(
+          std::min(std::abs(desired_acc[index]), joint_limits_[index].max_deceleration),
+          desired_acc[index]);
+      }
+      else if (joint_limits_[index].has_acceleration_limits)
+      {
+        desired_acc[index] = std::copysign(
+          std::min(std::abs(desired_acc[index]), joint_limits_[index].max_acceleration),
+          desired_acc[index]);
+      }
+
+      // Recompute velocity and position
+      if (has_vel_cmd)
+      {
+        desired_vel[index] =
+          current_joint_states.velocities[index] + desired_acc[index] * dt_seconds;
+      }
+      if (has_pos_cmd)
+        desired_pos[index] = current_joint_states.positions[index] +
+                             current_joint_states.velocities[index] * dt_seconds +
+                             0.5 * desired_acc[index] * dt_seconds * dt_seconds;
+
+      if (pos_limit_trig_jnts[index]) ostr << joint_names_[index] << " ";
+    }
+    ostr << "\b \b";  // erase last character
+    RCLCPP_WARN_STREAM_THROTTLE(
+      node_logging_itf_->get_logger(), *clock_, ROS_LOG_THROTTLE_PERIOD,
+      "Joint(s) [" << ostr.str().c_str()
+                   << "] would exceed position limits"
+                      " if continuing at current state, limiting all joints");
+  }
+
+  // display limitations
+
+  // if position limiting
+  if (
+    std::count_if(
+      pos_limit_trig_jnts.begin(), pos_limit_trig_jnts.end(), [](bool trig) { return trig; }) > 0)
+  {
+    std::ostringstream ostr;
+    for (size_t index = 0; index < number_of_joints_; ++index)
+    {
+      if (pos_limit_trig_jnts[index]) ostr << joint_names_[index] << " ";
+    }
+    ostr << "\b \b";  // erase last character
+    RCLCPP_WARN_STREAM_THROTTLE(
+      node_logging_itf_->get_logger(), *clock_, ROS_LOG_THROTTLE_PERIOD,
+      "Joint(s) [" << ostr.str().c_str() << "] would exceed position limits, limiting");
+  }
```

Comment:
Why is position limited joints are handled differently with a boolean, whereas other have a list of joints?, any specific reason?

Reply:
Awesome. Thanks